demos/greengrass_connectivity/aws_greengrass_discovery_demo.c

```diff
@@ -85,134 +142,347 @@ typedef struct
+    BackoffAlgorithm_InitializeParams( &xReconnectParams,
+                                       RETRY_BACKOFF_BASE_MS,
+                                       RETRY_MAX_BACKOFF_DELAY_MS,
+                                       RETRY_MAX_ATTEMPTS );
+
+    /* Attempt to connect to MQTT broker. If connection fails, retry after
+     * a timeout. Timeout value will exponentially increase till maximum
+     * attempts are reached.
+     */
+    do
     {
-        if( connectAttempt > 0 )
+        /* Establish a TLS session with the MQTT broker. This example connects to
+         * the MQTT broker as specified in pxHostAddressData->pcHostAddress and
+         * clientcredentialMQTT_BROKER_PORT at the top of this file. */
+        LogInfo( ( "Creating a TLS connection to %s:%u.",
+                   pxHostAddressData->pcHostAddress, clientcredentialMQTT_BROKER_PORT ) );
+        /* Attempt to create a mutually authenticated TLS connection. */
+        xNetworkStatus = SecureSocketsTransport_Connect( pxNetworkContext,
+                                                         &xServerInfo,
+                                                         &xSocketsConfig );
+
+        if( xNetworkStatus != TRANSPORT_SOCKET_STATUS_SUCCESS )
+        {
+            LogWarn( ( "Connection to the broker failed. Attempting connection retry after backoff delay." ) );
+
+            /* As the connection attempt failed, we will retry the connection after an
+             * exponential backoff with jitter delay. */
+
+            /* Calculate the backoff period for the next retry attempt and perform the wait operation. */
+            xBackoffStatus = prvBackoffForRetry( &xReconnectParams );
+        }
+        else
         {
-            IotLogError( "Failed to establish MQTT connection, retrying in %d ms.",
-                         ggdDEMO_RETRY_WAIT_MS );
-            IotClock_SleepMs( ggdDEMO_RETRY_WAIT_MS );
+            xMqttStatus = prvCreateMQTTConnectionWithBroker( pxHostAddressData, pxMQTTContext, pxNetworkContext );
+
+            if( xMqttStatus != pdPASS )
+            {
+                /* Calculate the backoff period for the next retry attempt and perform the wait operation. */
+                xBackoffStatus = prvBackoffForRetry( &xReconnectParams );
+                /* Close the network connection.  */
+                xNetworkStatus = SecureSocketsTransport_Disconnect( pxNetworkContext );
+
+                if( xNetworkStatus != TRANSPORT_SOCKET_STATUS_SUCCESS )
+                {
+                    LogError( ( "SecureSocketsTransport_Disconnect() failed to close the network connection. "
+                                "StatusCode=%d.", ( int ) xNetworkStatus ) );
+                }
+            }
         }
+    } while ( ( xNetworkStatus != TRANSPORT_SOCKET_STATUS_SUCCESS || xMqttStatus != pdPASS ) && ( xBackoffStatus == pdPASS ) );
+
+    return ( xNetworkStatus == TRANSPORT_SOCKET_STATUS_SUCCESS && xMqttStatus == pdPASS ) ? pdPASS : pdFAIL;
```

Comment:
The first clause is unnecessary since the second implies the former, i.e. `xNetworkStatus == TRANSPORT_SOCKET_STATUS_SUCCESS` whenever `xMqttStatus == pdPASS`.
```suggestion
    } while ( ( xMqttStatus != pdPASS ) && ( xBackoffStatus == pdPASS ) );

    return xMqttStatus;
```